cypress/src/integration/plugins/techdocs.spec.ts

```diff
@@ -0,0 +1,221 @@
+/*
+ * Copyright 2020 Spotify AB
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/// <reference types="cypress" />
+import 'os';
+
+describe('TechDocs', () => {
+  beforeEach(() => {
+    cy.loginAsGuest();
+    cy.mockTechDocsCSS();
+    cy.mockSockJSNode();
+    cy.interceptTechDocsAPICalls();
+  });
+
+  describe('Navigating to TechDocs', () => {
+    it('should navigate to the TechDocs home page via the primary navigation bar', () => {
+      cy.visit('/');
+      cy.get('[data-testid="sidebar-root"]')
+        .get('div')
+        .get('a[href="/docs"]')
+        .click();
+
+      cy.contains('Documentation');
+    });
+
+    it('should navigate to the TechDocs home page from the URL', () => {
+      cy.visit('/docs');
+      cy.contains('Documentation');
+    });
+
+    it('should navigate to a specific TechDocs entity from the "Overview" tab', () => {
+      cy.visit('/docs');
+      cy.get('[data-testid="read_docs"]').eq(0).click();
+
+      cy.location().should(loc => {
+        expect(loc.pathname).to.eq('/docs/default/Component/backstage');
+      });
+    });
+
+    it('should navigate to a specific TechDocs entity from the "Owned documents" tab', () => {
+      cy.visit('/docs');
+      cy.get('[data-testid="header-tab-1"]').click();
+      cy.get('[value="backstage"] > div > a').click();
+
+      cy.location().should(loc => {
+        expect(loc.pathname).to.eq('/docs/default/Component/backstage');
+      });
+    });
+
+    it('should navigate to a specific TechDocs entity page from a URL', () => {
+      cy.visit('/docs/default/Component/backstage');
+      cy.waitHomePage();
+
+      cy.contains('Backstage');
+      cy.contains(
+        'Main documentation for Backstage features and platform APIs',
+      );
+      cy.getTechDocsShadowRoot().contains(
+        'The Backstage documentation is available at',
+      );
+    });
+
+    it('should navigate to a specific TechDocs section from a URL', () => {
+      cy.visit('/docs/default/Component/backstage/overview/roadmap');
+      cy.waitRoadmapPage();
+
+      cy.window().its('scrollY').should('equal', 0);
+
+      cy.getTechDocsShadowRoot().within(() => {
```

Comment:
I kind of went back and forth in my head about the benefits/drawsbacks of relying on this documentation for the E2E tests, especially specific text on specific pages.

**For example**: It's possible for someone to make a change to this roadmap docs page in a branch...  And the E2E tests would pass (because it's loading docs from `master` on GitHub).  Once merged into into master, the E2E tests would start failing (because master was updated).

I wonder if it would be worth it to add a location for a separate fixture entity (in `tugboat.app-config.production.yaml`) whose `techdocs-ref` annotation used `dir:` instead of `url:`.  ...That way, the documentation used in E2E tests would always be loaded from the current branch, rather than what's in master.

...Of course the `dir` style of annotation is deprecated, but maybe we could save solving that for when we fully remove it.

What do you think?

Reply:
Yes, it makes sense to me too.

I don't know if this is what you meant in your comment but it would also be more relevant to make the e2e tests based on documentation that will not change in the future and that will be used only for these tests.

Reply:
Yeah, adding a specific set of docs to act as test fixture sounds like a plan.

I was thinking it would also be beneficial for the docs to be read locally, rather than via URL, so that you could make changes to the fixture doc and the E2E tests in the same PR.  e.g.

```diff
# .tugboat/tugboat.app-config.production.yaml

backend:
  baseUrl: ${TUGBOAT_DEFAULT_SERVICE_URL}
  cors:
    origin: ${TUGBOAT_DEFAULT_SERVICE_URL}
    methods: [GET, POST, PUT, DELETE]
    credentials: true

+catalog:
+  locations:
+    - type: file
+      target: ../path/to/techdocs/e2e-fixture.catalog.info.yaml
```

With a corresponding info yaml like...

```yaml
# e2e-fixture.catalog.info.yaml

apiVersion: backstage.io/v1alpha1
kind: Component
metadata:
  name: techdocs-e2e-fixture
  description: Used for end-to-end tests of TechDocs in Backstage.
  annotations:
    backstage.io/techdocs-ref: dir:./
```

Where `backstage.io/techdocs-ref: dir:./` is used instead of `backstage.io/techdocs-ref: url: https://github.com/etc`.

Reply:
Yeah, it sounds great!
I will make a custom set of docs to test the different TechDocs interactions.

A question that comes to my mind is about how to test this locally, outside of tugboat?
Should we add another command like `start-e2e-backend` that launch the backend with the correct set of docs?

Reply:
Ah, actually [it looks like](https://discord.com/channels/687207715902193673/687207715902193679/846659061582659624) we're not even running tugboat builds automatically right now.  So maybe it's fine to just put it in `app-config.yaml` for now.

\cc @freben ^

Reply:
Yeah tbh I don't see that adding those fixtures in the main `app-config.yaml` would hurt, so go ahead :grin: Especially compared to adding additional commands